core/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -279,12 +279,15 @@ private void processIndexShardSnapshots(ClusterChangedEvent event) {
                     SnapshotShards snapshotShards = shardSnapshots.get(entry.snapshot());
                     if (snapshotShards != null) {
                         for (ObjectObjectCursor<ShardId, ShardSnapshotStatus> shard : entry.shards()) {
-                            IndexShardSnapshotStatus snapshotStatus = snapshotShards.shards.get(shard.key);
+
+                            final IndexShardSnapshotStatus snapshotStatus = snapshotShards.shards.get(shard.key);
                             if (snapshotStatus != null) {
-                                switch (snapshotStatus.stage()) {
+                                final IndexShardSnapshotStatus.Copy lastSnapshotStatus = snapshotStatus.asCopy();
+                                final Stage stage = lastSnapshotStatus.getStage();
+                                switch (stage) {
                                     case INIT:
                                     case STARTED:
-                                        snapshotStatus.abort();
+                                        snapshotStatus.moveToAborted("snapshot has been aborted");
```

Comment:
same thing here. We could just call `abortIfNotCompleted`, and then take a snapshot and then act based on FINALIZE, DONE, FAILURE